lib/segment/src/index/field_index/mmap_point_to_values.rs

```diff
@@ -0,0 +1,385 @@
+    ) -> OperationResult<Self> {
+        // calculate file size
+        let header_size = std::mem::size_of::<Header>();
+        let points_count = iter
+            .clone()
+            .map(|(point_id, _)| (point_id + 1) as usize)
+            .max()
+            .unwrap_or(0);
+        let ranges_size = points_count * std::mem::size_of::<MmapRange>();
+        let values_size = iter
+            .clone()
+            .map(|v| v.1.iter().map(|v| v.mmaped_size()).sum::<usize>())
+            .sum::<usize>();
+        let file_size = header_size + ranges_size + values_size;
+
+        // create new file and mmap
+        let file_name = path.join(POINT_TO_VALUES_PATH);
+        create_and_ensure_length(&file_name, file_size)?;
+        let mut mmap = open_write_mmap(&file_name)?;
+
+        // fill mmap file data
+        let header = Header {
+            ranges_start: header_size as u64,
+            points_count: points_count as u64,
+        };
+        header
+            .write_to_prefix(mmap.as_mut())
+            .ok_or_else(|| OperationError::service_error(NOT_ENOUGHT_BYTES_ERROR_MESSAGE))?;
+
+        // counter for values offset
+        let mut point_values_offset = header_size + ranges_size;
+        for (point_id, values) in iter {
+            let range = MmapRange {
+                start: point_values_offset as u64,
+                count: values.len() as u64,
+            };
+            mmap.get_mut(header_size + point_id as usize * std::mem::size_of::<MmapRange>()..)
+                .and_then(|bytes| range.write_to_prefix(bytes))
+                .ok_or_else(|| OperationError::service_error(NOT_ENOUGHT_BYTES_ERROR_MESSAGE))?;
+
+            for value in values {
+                let bytes = mmap.get_mut(point_values_offset..).ok_or_else(|| {
+                    OperationError::service_error(NOT_ENOUGHT_BYTES_ERROR_MESSAGE)
+                })?;
+                value.write_to_mmap(bytes)?;
+                point_values_offset += value.mmaped_size();
+            }
+        }
+
+        mmap.flush()?;
+        Ok(Self {
+            file_name,
+            mmap: mmap.make_read_only()?,
+            header,
+        })
+    }
+
+    pub fn open(path: &Path) -> OperationResult<Self> {
+        let file_name = path.join(POINT_TO_VALUES_PATH);
+        let mmap = open_write_mmap(&file_name)?;
+        let header =
+            Header::read_from_prefix(mmap.as_ref()).ok_or(OperationError::InconsistentStorage {
+                description: NOT_ENOUGHT_BYTES_ERROR_MESSAGE.to_owned(),
+            })?;
+
+        Ok(Self {
+            file_name,
+            mmap: mmap.make_read_only()?,
+            header,
+        })
+    }
+
+    fn files(&self) -> Vec<PathBuf> {
+        vec![self.file_name.clone()]
+    }
+
+    pub fn get_values<'a, T: MmapValue<'a>>(
+        &'a self,
+        point_id: PointOffsetType,
+    ) -> Box<dyn Iterator<Item = OperationResult<T>> + 'a> {
```

Comment:
Would it be possible to return concrete iterator type instead?

This is supposed to be a very hot function

Reply:
We need another approach at all. I have already integrated this iterator here:
https://github.com/qdrant/qdrant/pull/4644
But it seems we have perf degradation

Reply:
Fixed, changes from https://github.com/qdrant/qdrant/pull/4665 were applied here